Userland/xargs.cpp

```diff
@@ -0,0 +1,304 @@
+{
+    bool fail = false;
+
+    for (;;) {
+        char* item = nullptr;
+        size_t buffer_size = 0;
+
+        auto item_size = getdelim(&item, &buffer_size, entry_separator, fp);
+
+        if (item_size < 0) {
+            // getdelim() will return -1 and set errno to 0 on EOF.
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        Decision decision;
+        do {
+            decision = callback(item);
+            if (decision == Stop) {
+                free(item);
+                return true;
+            }
+        } while (decision == Unget);
+
+        free(item);
+    }
+
+    return fail;
+}
+
+bool run_command(Vector<char*>& child_argv, bool verbose, bool is_stdin, int devnull_fd)
+{
+    child_argv.append(nullptr);
+
+    if (verbose) {
+        StringBuilder builder;
+        builder.join(" ", child_argv);
+        fprintf(stderr, "xargs: %s\n", builder.to_string().characters());
+        fflush(stderr);
+    }
+
+    auto pid = fork();
+    if (pid < 0) {
+        perror("fork");
+        return false;
+    }
+
+    if (pid == 0) {
+        if (is_stdin)
+            dup2(devnull_fd, STDIN_FILENO);
+
+        execvp(child_argv[0], child_argv.data());
+        exit(1);
+    }
+
+    for (auto* ptr : child_argv)
+        free(ptr);
+
+    child_argv.clear();
+
+    int wstatus = 0;
+    if (waitpid(pid, &wstatus, 0) < 0) {
+        perror("waitpid");
+        return false;
+    }
+
+    if (WIFEXITED(wstatus)) {
+        if (WEXITSTATUS(wstatus) != 0)
+            return false;
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+ParsedInitialArguments::ParsedInitialArguments(Vector<const char*>& arguments, const StringView& placeholder)
+{
+    m_all_parts.ensure_capacity(arguments.size());
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        if (placeholder.is_empty()) {
+            parts.append(arg);
+        } else {
+            auto maybe_placeholder_index = arg.find_first_of(placeholder);
+            size_t last_placeholder_end_offset = 0;
+            // Split the argument into segments delimited by the placeholder.
+            while (maybe_placeholder_index.has_value()) {
+                some_argument_has_replacement = true;
+                auto placeholder_end_offset = maybe_placeholder_index.value() + placeholder.length();
+                auto part = arg.substring_view(last_placeholder_end_offset, placeholder_end_offset - placeholder.length());
+                parts.append(part);
+                last_placeholder_end_offset += placeholder_end_offset;
+                maybe_placeholder_index = arg.substring_view(last_placeholder_end_offset, arg.length() - last_placeholder_end_offset).find_first_of(placeholder);
```

Comment:
And if you do that, I think you'll be able to use `substring_view_starting_after_substring()` here nicely after all, and get rid of even more index math.